src/content/app/genome-browser/components/track-panel/components/track-panel-list/track-panel-items/TrackPanelGene.tsx

```diff
@@ -67,75 +59,76 @@ const TrackPanelGene = (props: TrackPanelGeneProps) => {
   );
   const visibleTranscriptIds = useAppSelector((state) => {
     const genomeTrackStates = getBrowserActiveGenomeTrackStates(state);
-    return genomeTrackStates?.objectTracks?.[focusObjectId]?.transcripts ?? [];
+    return (
+      genomeTrackStates?.objectTracks?.[focusObjectId]?.transcripts ?? null
+    );
   });
-  const { toggleTrack, updateFocusGeneTranscripts, genomeBrowser } =
-    useGenomeBrowser();
-  const dispatch = useAppDispatch();
 
-  const allTranscriptsInGene = currentData?.gene.transcripts ?? [];
-
-  useEffect(() => {
-    const subscription = genomeBrowser?.subscribe(
-      IncomingActionType.VISIBLE_TRANSCRIPTS,
-      (action: ReportVisibleTranscriptsAction) => {
-        setVisibleTranscriptIds(action.payload.transcript_ids);
-      }
-    );
+  const { setFocusGene, updateFocusGeneTranscripts } = useGenomeBrowser();
+  const dispatch = useAppDispatch();
 
-    return () => subscription?.unsubscribe();
-  }, [genomeBrowser]);
+  if (!currentData) {
+    return null;
+  }
 
-  useEffect(() => {
-    updateObjectTrackStatus(trackStatus);
-  }, [genomeId, focusObjectId, trackStatus]);
+  const { gene } = currentData;
 
-  // set status of all transcripts based on the saved redux state after loading component
-  useEffect(() => {
-    if (allTranscriptsInGene?.length) {
+  const sortedTranscripts = defaultSort(gene.transcripts);
+  const visibleSortedTranscripts = isCollapsed
+    ? [sortedTranscripts[0]]
+    : sortedTranscripts;
+
+  const geneVisibilityStatus = !visibleTranscriptIds?.length
+    ? Status.UNSELECTED
+    : visibleTranscriptIds.length === gene.transcripts.length
+    ? Status.SELECTED
+    : Status.PARTIALLY_SELECTED;
+
+  const onGeneVisibilityChange = () => {
+    if (geneVisibilityStatus === Status.PARTIALLY_SELECTED) {
+      // show all transcripts
+      const visibleTranscriptIds = sortedTranscripts.map(
+        (transcript) => transcript.stable_id
+      );
       updateFocusGeneTranscripts(visibleTranscriptIds);
+      return;
     }
-  }, [allTranscriptsInGene]);
 
-  const setVisibleTranscriptIds = (transcriptIds: string[]) => {
-    dispatch(
-      updateObjectTrackStates({
-        status: transcriptIds.length ? Status.SELECTED : Status.UNSELECTED,
-        transcriptIds
-      })
-    );
-  };
+    const newStatus =
+      trackStatus === Status.SELECTED ? Status.UNSELECTED : Status.SELECTED;
 
-  const updateObjectTrackStatus = (newStatus?: Status) => {
-    if (!newStatus) {
-      newStatus =
-        trackStatus === Status.SELECTED ? Status.UNSELECTED : Status.SELECTED;
+    if (newStatus === Status.SELECTED) {
+      setFocusGene(focusObjectId);
+      const visibleTranscriptIds = sortedTranscripts.map(
+        (transcript) => transcript.stable_id
+      );
```

Comment:
Isn't this variable the same as the one at line 90?

Reply:
Extracted this into a function. Not sure whether that's an improvement 😃

Reply:
That definitely looks better 😀